runtime/compiler/runtime/J9CodeCacheManager.cpp

```diff
@@ -355,18 +359,46 @@ J9::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
 
    void *defaultEndAddress = vmemParams.endAddress;
 
+   uintptr_t someJitLibraryAddress = self()->getSomeJitLibraryAddress();
+
+#if defined(TR_TARGET_POWER)
+   size_t alignment = 64 * 1024;
+#elif (defined(LINUX) && defined(TR_TARGET_X86))
+   size_t alignment = 2 * 1024 * 1024;
+#elif (defined(TR_TARGET_S390))
+   size_t alignment = 1024 * 1024;
+#else
+   size_t alignment = pageSizes[0];
+#endif
+   if (largeCodePageSize > 0)
+      alignment = largeCodePageSize;
+
    if (preferredStartAddress)
       {
-   // J9PORT_VMEM_ADDRESS_HINT is only supported for default pages and linux
+      vmemParams.options |= J9PORT_VMEM_STRICT_ADDRESS; // if we cannot allocate a block in the desired range, return NULL
+      vmemParams.alignmentInBytes = alignment;
 #if defined(LINUX)
-      if (largeCodePageSize == 0)
+      vmemParams.options |= J9PORT_VMEM_ALLOC_QUICK; // Use smaps for guidance
+      // If jit library address is greater than MAX_DISTANCE (~2GB),
+      // then we search in the addresses before the jit library address.
+      // If jit library address is less than MAX_DISTANCE, (which means we
+      // do not have the full MAX_DISTANCE range before the jit library),
+      // then we search in the addresses after the jit library address
+      if (someJitLibraryAddress > MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE)
          {
-         vmemParams.options |= J9PORT_VMEM_ADDRESS_HINT;
+         // Align the startAddress to page boundary
+         vmemParams.startAddress = (void *)OMR::align((size_t)(someJitLibraryAddress - MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE), alignment);
+         vmemParams.endAddress = (void *)(((UDATA)preferredStartAddress) + SAFE_DISTANCE_REPOSITORY_JITLIBRARY);
          }
+      else
+         {
+         vmemParams.startAddress = (void *)OMR::align((size_t)(someJitLibraryAddress + SAFE_DISTANCE_REPOSITORY_JITLIBRARY), alignment);
+         vmemParams.endAddress = (void *)(someJitLibraryAddress + MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE - segmentSize);
```

Comment:
`MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE` is technically less than 2GB, is that ok? Feels like it should be
```
someJitLibraryAddress + SAFE_DISTANCE_REPOSITORY_JITLIBRARY 
 + MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE - segmentSize
```
right?

Reply:
Also why subtract `segmentSize`?

Reply:
`MAX_DISTANCE_NEAR_JITLIBRARY_TO_AVOID_TRAMPOLINE` is 2GB (maximum target for a branch on x86) minus some fudge value that accounts for the maximum size of the JIT dll. We should consider the fact that `someJitLibraryAddress` could be at the very beginning of at the very end of the JIT dll. Also, the JIT dll may increase in the future, so the fudge value was picked to be 64 MB.

This case corresponds to the situation where the JITdll is loaded in low addresses and we don't have a full (2GB-64MB) before it (unlikely in practical terms). In this case we want to allocate the code cache in the (2GB-64MB) space after `someJITLibrary` address.
Hence startAddress is `someJITdllAddress + 64 MB` and the end address is `someJItDllAddress + (2GB-64MB) -repoSize`. We need to substract repoSize because what we search is the beginning of our repository segment and we want the entire repository segment to benefit from avoiding the trampolines.

Reply:
Ah I see, `segmentSize` here is the size of the repo